File: scripts/update_route_endpoints.py
import json
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

ROUTES_FILE = Path("data/routes.json")
TFL_URL = "https://api.tfl.gov.uk/Line/Mode/bus/Route"


def fetch_tfl_routes() -> list[dict]:
    params = {}
    api_key = os.environ.get("TFL_API_KEY", "").strip()
    if api_key:
        params["app_key"] = api_key

    url = TFL_URL
    if params:
        url = f"{url}?{urllib.parse.urlencode(params)}"

    request = urllib.request.Request(
        url,
        headers={
            "Accept": "application/json",
            "User-Agent": "RFL-Route-Database-Updater/1.0",
        },
    )

    for attempt in range(4):
        try:
            with urllib.request.urlopen(request, timeout=60) as response:
                return json.load(response)
        except urllib.error.HTTPError as error:
            if error.code == 429 and attempt < 3:
                retry_after = error.headers.get("Retry-After")
                delay = int(retry_after) if retry_after and retry_after.isdigit() else 15 * (attempt + 1)
                print(f"TfL returned 429. Retrying in {delay} seconds...", file=sys.stderr)
                time.sleep(delay)
                continue
            raise

    raise RuntimeError("TfL route data could not be downloaded")


def endpoint_pair(line: dict) -> tuple[str, str] | None:
    sections = line.get("routeSections") or []
    if not sections:
        return None

    outbound = [
        section
        for section in sections
        if str(section.get("direction", "")).lower() == "outbound"
        and section.get("originationName")
        and section.get("destinationName")
    ]
    candidates = outbound or [
        section
        for section in sections
        if section.get("originationName") and section.get("destinationName")
    ]
    if not candidates:
        return None

    section = candidates[0]
    return str(section["originationName"]).strip(), str(section["destinationName"]).strip()


def main() -> None:
    database = json.loads(ROUTES_FILE.read_text(encoding="utf-8"))
    tfl_lines = fetch_tfl_routes()

    endpoints = {}
    for line in tfl_lines:
        line_id = str(line.get("id", "")).strip().lower()
        pair = endpoint_pair(line)
        if line_id and pair:
            endpoints[line_id] = pair

    updated = 0
    unchanged = 0
    missing = []

    for route in database.get("current", []):
        route_id = str(route.get("Route", "")).strip()
        pair = endpoints.get(route_id.lower())
        if not pair:
            missing.append(route_id)
            continue

        start, destination = pair
        if route.get("Start") != start or route.get("Destination") != destination:
            print(
                f"{route_id}: {route.get('Start')} -> {route.get('Destination')} "
                f"becomes {start} -> {destination}"
            )
            route["Start"] = start
            route["Destination"] = destination
            updated += 1
        else:
            unchanged += 1

    ROUTES_FILE.write_text(
        json.dumps(database, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )

    print(f"Updated {updated} routes; {unchanged} already matched TfL.")
    if missing:
        print(f"TfL did not return endpoints for {len(missing)} routes: {', '.join(missing)}")


if __name__ == "__main__":
    main()

File: scripts/update-route-endpoints.mjs
import { readFile, writeFile } from "node:fs/promises";

const databasePath = new URL("../data/routes.json", import.meta.url);
const apiUrl = new URL("https://api.tfl.gov.uk/Line/Mode/bus/Route");
apiUrl.searchParams.set("serviceTypes", "Regular");

if (process.env.TFL_API_KEY) {
  apiUrl.searchParams.set("app_key", process.env.TFL_API_KEY);
}

const response = await fetch(apiUrl, {
  headers: { Accept: "application/json" }
});

if (!response.ok) {
  throw new Error(`TfL returned ${response.status} ${response.statusText}`);
}

const lines = await response.json();
const database = JSON.parse(await readFile(databasePath, "utf8"));

function normalise(value) {
  return String(value ?? "").trim().toLowerCase();
}

function routeId(line) {
  return normalise(line.id || line.name);
}

function endpointPair(line) {
  const sections = Array.isArray(line.routeSections) ? line.routeSections : [];
  const usable = sections.filter(section =>
    String(section.originationName ?? "").trim() &&
    String(section.destinationName ?? "").trim()
  );

  const outbound = usable.filter(section => normalise(section.direction) === "outbound");
  const selected = outbound[0] || usable[0];

  if (!selected) return null;
  return {
    start: String(selected.originationName).trim(),
    destination: String(selected.destinationName).trim()
  };
}

const tflRoutes = new Map();
for (const line of lines) {
  const pair = endpointPair(line);
  if (pair) tflRoutes.set(routeId(line), pair);
}

let updated = 0;
let unchanged = 0;
let missing = 0;

for (const route of database.current || []) {
  const pair = tflRoutes.get(normalise(route.Route));
  if (!pair) {
    missing += 1;
    console.warn(`No TfL endpoint found for route ${route.Route}`);
    continue;
  }

  if (route.Start === pair.start && route.Destination === pair.destination) {
    unchanged += 1;
    continue;
  }

  route.Start = pair.start;
  route.Destination = pair.destination;
  updated += 1;
  console.log(`Updated ${route.Route}: ${pair.start} -> ${pair.destination}`);
}

await writeFile(databasePath, `${JSON.stringify(database, null, 2)}\n`, "utf8");
console.log(`Finished: ${updated} updated, ${unchanged} unchanged, ${missing} not found.`);
